incubator/virtualcluster/pkg/syncer/controllers/pod/checker.go

```diff
@@ -0,0 +1,137 @@
+	wait.Until(c.checkPods, c.periodCheckerPeriod, stopCh)
+}
+
+// checkPods checks to see if pods in super master informer cache and tenant master
+// keep consistency.
+func (c *controller) checkPods() {
+	clusterNames := c.multiClusterPodController.GetClusterNames()
+	if len(clusterNames) == 0 {
+		klog.Infof("tenant masters has no clusters, give up period checker")
+		return
+	}
+
+	for _, clusterName := range clusterNames {
+		cluster := c.multiClusterPodController.GetCluster(clusterName)
+		if cluster == nil {
+			klog.Errorf("failed to locate cluster %s", clusterName)
+			continue
+		}
+
+		c.checkPodsOfCluster(clusterName, cluster)
+	}
+
+	pPods, err := c.podLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("error listing pods from super master informer cache: %v", err)
+		return
+	}
+
+	for _, pPod := range pPods {
+		clusterName, vNamespace := conversion.GetOwner(pPod)
+		if len(clusterName) == 0 || len(vNamespace) == 0 {
+			continue
+		}
+		vPodObj, err := c.multiClusterPodController.Get(clusterName, vNamespace, pPod.Name)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				if pPod.DeletionTimestamp != nil {
+					// pPod is under deletion, waiting for UWS bock populate the pod status.
+					continue
+				}
+				c.multiClusterPodController.RequeueObject(clusterName, vPodObj, reconciler.DeleteEvent)
+			}
+		}
+		vPod := vPodObj.(*v1.Pod)
+
+		// pod has been updated by super master
+		if !equality.Semantic.DeepEqual(vPod.Status, pPod.Status) {
+			klog.Warningf("status of pod %v/%v diff in super&tenant master", pPod.Namespace, pPod.Name)
+		}
+	}
+}
+
+// checkPodsOfCluster checks to see if pods in specific cluster keeps consistency.
+func (c *controller) checkPodsOfCluster(clusterName string, cluster ctrl.ClusterInterface) {
+	clusterInformerCache, err := cluster.GetCache()
+	if err != nil {
+		klog.Errorf("failed to get informer cache for cluster %s", clusterName)
+		return
+	}
+	podList := &v1.PodList{}
+	err = clusterInformerCache.List(context.TODO(), podList)
+	if err != nil {
+		klog.Errorf("error listing pods from cluster informer cache: %v", err)
+		return
+	}
+
+	klog.Infof("check pods consistency in cluster %s", clusterName)
+	for _, vPod := range podList.Items {
+		targetNamespace := conversion.ToSuperMasterNamespace(clusterName, vPod.Namespace)
+		pPod, err := c.podLister.Pods(targetNamespace).Get(vPod.Name)
+		if err == nil {
+			updatedPod := conversion.CheckPodEquality(pPod, &vPod)
+			if updatedPod != nil {
+				klog.Warningf("spec of pod %v/%v diff in super&tenant master", vPod.Namespace, vPod.Name)
+			}
+		} else if errors.IsNotFound(err) {
+			// pPod not found and vPod is under deletion, we need to delete vPod manually
+			if vPod.DeletionTimestamp != nil {
+				clusterClient, err := cluster.GetClient()
+				if err != nil {
+					klog.Errorf("error getting cluster %s clientset: %v", err)
+					continue
+				}
+				deleteOptions := metav1.NewDeleteOptions(*vPod.DeletionGracePeriodSeconds)
```

Comment:
DeletionGracePeriodSeconds should be 0 here, or you cannot actually delete the vPod

Reply:
yep, since pPod not found in super master, we can force delete vPod。